src/expr/lazy_proof_chain.cpp

```diff
@@ -107,7 +107,10 @@ std::shared_ptr<ProofNode> LazyCDProofChain::getProofFor(Node fact)
           << "\n";
       // retrieve free assumptions and their respective proof nodes
       std::map<Node, std::vector<std::shared_ptr<ProofNode>>> famap;
-      expr::getFreeAssumptionsMap(curPfn, famap);
+      if (curPfn != nullptr)
+      {
+        expr::getFreeAssumptionsMap(curPfn, famap);
+      }
```

Comment:
let's move this test to before checking `!rec`. If `curPfn` is `null` we mark cur as post-visited and skip. Only if it is not null we add it to `toConnect`. This will also remove the necessity of the test you added in the end.

Reply:
Is 2459f857ddca0318cc8cd10ae271d40d72509d77 what you had in mind?

Reply:
Almost. Just need to also do `visited[cur] = true;` when `curPfn` is null.

Reply:
Actually id merge this with the `rec` condition. This also means that we cache toConnect[cur] being nullptr.

Reply:
Currently it does not cache null. When `rec` is false though we do cache. I'd not like to cache null, as it'd require another test after the visiting.

Reply:
We already have that test actually (https://github.com/CVC4/CVC4/blob/proof-new/src/expr/lazy_proof_chain.cpp#L225).  I added it.  I'd prefer caching nullptr here since it makes the invariant of the data structure cleaner.